DOC: update SciPy 1.16.0 release notes

tylerjereddy opened the pull request:
* Update the SciPy `1.16.0` release notes prior to branching. I'm a bit behind schedule--we had 50 milestone-labeled PRs at the start of the week are we now hovering around 25. I'll probably have to get tougher about bumping milestones for things that aren't 1-2 days away from crossing the line.

[ci skip] [skip ci]

TODO:

- [ ] `.mailmap` updates
- [x] warnings in the doc build
- [x] manually scanning through release note PRs with `enhancement` labels (Matt will be doing this for `stats`)
- [x] check the release notes Wiki again for updates--initial transcription was performed on May 15/2025; checked again on May 17th
- [x] release notes highlights section
- [x] gradually reactivate the CI here, when ready
- [x] Remove unused/empty (sub)sections in the release notes
- [x] re-run author/list scripts again just before branching

tylerjereddy committed:
DOC: update SciPy 1.16.0 release notes

* Update the SciPy `1.16.0` release notes
prior to branching.

[ci skip] [skip ci]

mdhaber committed:
DOC: update 1.16.0 release notes

tylerjereddy commented:
I merged a fork-based PR from Matt to update some sections just now--I'll try to add a few inline comments carried over from discussion points over there.

tylerjereddy commented on doc/source/release/1.16.0-notes.rst:
```diff
@@ -88,43 +124,832 @@ New features
+in `scipy.stats`, `scipy.special`, `scipy.optimize`, and `scipy.constants`
+now provide tables documenting compatible array and device types as well as
+support for lazy arrays and JIT compilation. New features with support and
+old features with support added for SciPy 1.16.0 include:
```
There was discussion in https://github.com/tylerjereddy/scipy/pull/83 about whether we should re-list parts of SciPy that previously had array API support, but didn't have i.e., JAX or Dask support.

Or whether it would be preferable to only include parts of SciPy for which basic array API support previously wasn't present at all.

There are some verbosity tradeoffs to consider perhaps.

rgommers replied:
I'd not go overboard, but add something like: "In addition, some functionality that already supported other array libraries in previous releases extended that support - most commonly by adding JAX and Dask support in addition to PyTorch and CuPy."

tylerjereddy commented on doc/source/release/1.16.0-notes.rst:
```diff
@@ -41,37 +41,69 @@ New features
 
+``scipy.interpolate`` improvements
+==================================
+- Added batch support to `scipy.interpolate.make_smoothing_spline`
```
I think `interpolate` improvement section title here is duplicated from above, unless I'm just misreading this somehow. Can fuse easily enough.

mdhaber replied:
Ooops https://github.com/scipy/scipy/pull/22991/files#r2093784293

mdhaber commented on doc/source/release/1.16.0-notes.rst:
```diff
@@ -41,37 +41,69 @@ New features
 ====================================
 
-
 ``scipy.interpolate`` improvements
```
```suggestion
``scipy.integrate`` improvements
```

tylerjereddy committed:
DOC: PR 22991 revisions

* Update `.mailmap` based on discussion at
gh-23004.

[ci skip] [skip ci]

tylerjereddy commented on .mailmap:
```diff
@@ -207,7 +207,8 @@ Franziska Horn <cod3licious@users.noreply.github.com> cod3licious <cod3licious@u
 G Young <gfyoung17@gmail.com> gfyoung <gfyoung@mit.edu>
-Gagandeep Singh <gdp.1807@gmail.com> czgdp1807 <gdp.1807@gmail.com>
+ਗਗਨਦੀਪ ਸਿੰਘ (Gagandeep Singh) <gdp.1807@gmail.com> czgdp1807 <gdp.1807@gmail.com>
+ਗਗਨਦੀਪ ਸਿੰਘ (Gagandeep Singh) <gdp.1807@gmail.com> Gagandeep Singh <gdp.1807@gmail.com>
```
based on Guido's observation at gh-23004

tylerjereddy commented:
I'm working on this a bit locally and not ready to push up yet--I'll continue tomorrow (May 17th). It looks like the team has done some heavy lifting to get the PR list down to 9 left.

dschmitz89 commented on doc/source/release/1.16.0-notes.rst:
```diff
@@ -88,43 +124,831 @@ New features
+- For ``Cython>=3.1.0b1``, SciPy now uses the new
+  ``cython --generate-shared`` functionality, which reduces the total size of
+  SciPy's wheels and on-disk installations significantly.
+- The output of `scipy.stats.wrapcauchy.rvs` is now mapped to the unit circle.
```
```suggestion
- The output of `scipy.stats.wrapcauchy.rvs` is now mapped to the unit circle between 0 and 2 * pi.
```

rgommers commented:
>  It looks like the team has done some heavy lifting to get the PR list down to 9 left.

Down to 4 PRs now, I'll have a look at those and will see if I can write a few things for the release notes too.

lucascolley commented on doc/source/release/1.16.0-notes.rst:
```diff
@@ -88,43 +124,831 @@ New features
+  same shape (ignoring ``axis``) before performing the calculation.
 
 *************
 Other changes
```
I think we should add a notice to distributors along the lines of https://github.com/scipy/scipy/issues/17751#issuecomment-2884399570

rgommers replied:
Here is proposed text:

- The vendored Qhull library was upgraded from version 2019.1 to 2020.2
- A new build option `-Duse-system-libraries` has been added. It allows opting in to using system libraries instead of using vendored sources. Currently Boost.Math and Qhull are supported as system build dependencies.

DietBru commented on doc/source/release/1.16.0-notes.rst:
```diff
@@ -41,37 +41,69 @@ New features
+-
+- A new function ``firwin_2d`` for the creation of a 2-D FIR Filter using the
+  1-D window method was added.
 
```
```suggestion
- A new function `scipy.signal.closest_STFT_dual_window` to calculate the `ShortTimeFFT` dual window of a given window closest to a desired dual window.
- A new classmethod `scipy.signal.ShortTimeFFT.from_win_equals_dual` to create a `ShortTimeFFT` instance where the window and its dual are equal up to a scaling factor. It allows to create short-time Fourier transforms which are unitary mappings.

```

rgommers commented:
Okay, all PRs except this one merged now, all signs on green:)

tylerjereddy commented:
excellent, many thanks, I'll push on the release notes today--I may ask for opinions on highlights (etc.) after I draft them; could refine after branching too

dschult committed:
sparse text for release notes [docs only]

tylerjereddy committed:
DOC: PR 22991 revisions

* Update the release notes for SciPy `1.16.0` after manually
examining the list of PRs with the `1.16.0` milestone label
and an `enhancement` label.

[ci skip] [skip ci]

tylerjereddy commented:
I completed the manual scan through of the 99 enhancement-labeled PRs with the `1.16.0` milestone and filled in any release notes gaps I noticed. The team did a pretty solid job before the manual scan--only about 60 lines of changes, far less than in the past.

tylerjereddy committed:
DOC: PR 22991 revisions

* Purge unused/empty (sub)sections in the
SciPy `1.16.0` release notes.

[ci skip] [skip ci]

tylerjereddy committed:
DOC: PR 22991 revisions

* Update the author/issue/PR lists as we get
closer to the branching time.

[ci skip] [skip ci]

tylerjereddy committed:
DOC: PR 22991 revisions

* Update the "Highlights" section of the SciPy
`1.16.0` release notes.

* Misc. minor fixes/edits elsewhere in the SciPy
`1.16.0` release notes.

[docs only]

tylerjereddy commented:
Alright, the docs build passed in CI and the rendered version of the `1.16.0` release notes looks "ok" there as well as locally.

Some things are not perfect yet: `.mailmap` needs quite a bit of work (even some core developers have duplicated names..), "Highlights" may be slightly too long now, and we may be missing a few deprecation mentions perhaps. Still, I think a bias toward action is the right call here--the project is quiet this evening so this is a great time to branch and those issues can be handled on the maintenance branch in the next day or two. Also, staying on schedule is probably a good idea.

rgommers commented on doc/source/release/1.16.0-notes.rst:
```diff
@@ -17,114 +17,1035 @@ run your code with ``python -Wd`` and check for ``DeprecationWarning`` s).
+- ``csgraph.dijkstra`` shortest_path is more efficient.
+- ``csgraph.yen`` has performance improvements.
+- Import of ``_lib.issparse`` allows checking for ``scipy.sparse`` object
+  without full import of ``scipy``.
```
This should be removed or rewritten I think. The private utility function isn't relevant to readers of release notes, so we should express it as the change in behavior for public functions (if relevant).

rgommers commented on doc/source/release/1.16.0-notes.rst:
```diff
@@ -17,114 +17,1035 @@ run your code with ``python -Wd`` and check for ``DeprecationWarning`` s).
 
 ``scipy.sparse`` improvements
 =============================
+- ``coo_array`` supports nD arrays using binary and reduction operations.
```
nit: we normally write this as n-D

rgommers commented on doc/source/release/1.16.0-notes.rst:
```diff
@@ -17,114 +17,1035 @@ run your code with ``python -Wd`` and check for ``DeprecationWarning`` s).
+- ``csgraph.yen`` has performance improvements.
+- Import of ``_lib.issparse`` allows checking for ``scipy.sparse`` object
+  without full import of ``scipy``.
+- add lazy loading of ``sparse.csgraph`` and ``sparse.linalg``.
```
I'd like to rephrase this.

rgommers commented on doc/source/release/1.16.0-notes.rst:
```diff
@@ -17,114 +17,1035 @@ run your code with ``python -Wd`` and check for ``DeprecationWarning`` s).
+  - work with existing discrete distributions and
+  - facilitate the creation of custom distributions
 
+  in the new random variable infrastructure.
```
nit: this should probably just be a single sentence rather than nested bullets.

rgommers commented on doc/source/release/1.16.0-notes.rst:
```diff
@@ -17,114 +17,1035 @@ run your code with ``python -Wd`` and check for ``DeprecationWarning`` s).
+  ``lm`` method the number of function calls used in Jacobian approximation
+  is no longer included in ``OptimizeResult.nfev``. This brings the behavior
+  of ``lm``, ``trf``, and ``dogbox`` into line.
+- For ``Cython>=3.1.0b1``, SciPy now uses the new
```
This should be `>=3.1.0` now.

rgommers commented on doc/source/release/1.16.0-notes.rst:
```diff
@@ -17,114 +17,1035 @@ run your code with ``python -Wd`` and check for ``DeprecationWarning`` s).
+- For ``Cython>=3.1.0b1``, SciPy now uses the new
+  ``cython --generate-shared`` functionality, which reduces the total size of
+  SciPy's wheels and on-disk installations significantly.
+- The output of `scipy.stats.wrapcauchy` ``rvs`` is now mapped to the unit circle
```
This needs a tweak, maybe "the rvs method of ..."

rgommers commented on doc/source/release/1.16.0-notes.rst:
```diff
@@ -17,114 +17,1035 @@ run your code with ``python -Wd`` and check for ``DeprecationWarning`` s).
+  dependencies.
+- The ``namedtuple``-like bunch objects returned by some SciPy functions
+  now have improved compatibility with the ``polars`` library.
+- The testsuite thread safety with free-threaded CPython has improved.
```
I'll rephrase and add to this. It was more than just the test suite -e.g.,  we finished making `sparse` and `special` thread-safe.

rgommers reviewed (commented):
Thanks Tyler. This looks like a great release!

Overall looks good. I only had time to review the full contents now, added some inline comments and in addition we should include:

- Introduction of `spin` as the new developer CLI (`dev.py` still present), with improved support for editable installs.
- Size reduction of the sdist by 50% and making its contents more accurate
- Under improved array API support: that we now have a GPU CI job
- New minimum versions of Clang, and for wheels minimum macOS version bump (gh-22504)
- SciPy no longer using an internal shared library (`sf_error_state` was removed)
- The creation of `xsf`

I'll open a PR to the 1.16.x branch for these.
